ts/lib/shortcuts.ts

```diff
@@ -0,0 +1,149 @@
+import * as tr from "./i18n";
+
+export type Modifier = "Control" | "Alt" | "Shift" | "Meta";
+
+const modifiers: Modifier[] = ["Control", "Alt", "Shift", "Meta"];
+
+// how modifiers are mapped
+const platformModifiers =
+    navigator.platform === "MacIntel"
+        ? ["Meta", "Alt", "Shift", "Control"]
+        : ["Control", "Alt", "Shift", "OS"];
+
+function splitKeyCombinationString(keyCombinationString: string): string[][] {
+    return keyCombinationString.split(", ").map((segment) => segment.split("+"));
+}
+
+function modifiersToPlatformString(modifiers: string[]): string {
+    const displayModifiers =
+        navigator.platform === "MacIntel"
+            ? ["^", "⌥", "⇧", "⌘"]
+            : [`${tr.keyboardCtrl()}+`, "Alt+", `${tr.keyboardShift()}+`, "Win+"];
+
+    let result = "";
+
+    for (const modifier of modifiers) {
+        result += displayModifiers[platformModifiers.indexOf(modifier)];
+    }
+
+    return result;
+}
+
+const alphabeticPrefix = "Key";
+const numericPrefix = "Digit";
+const keyToCharacterMap = {
+    Backslash: "\\",
+    Backquote: "`",
+    BracketLeft: "[",
+    BrackerRight: "]",
+    Quote: "'",
+    Semicolon: ";",
+    Minus: "-",
+    Equal: "=",
+    Comma: ",",
+    Period: ".",
+    Slash: "/",
+};
+
+function keyToPlatformString(key: string): string {
+    return key.startsWith(alphabeticPrefix)
+        ? key.slice(alphabeticPrefix.length)
+        : key.startsWith(numericPrefix)
+        ? key.slice(numericPrefix.length)
+        : Object.prototype.hasOwnProperty.call(keyToCharacterMap, key)
+        ? keyToCharacterMap[key]
+        : key;
+}
+
+function toPlatformString(modifiersAndKey: string[]): string {
+    return `${modifiersToPlatformString(
+        modifiersAndKey.slice(0, -1)
+    )}${keyToPlatformString(modifiersAndKey[modifiersAndKey.length - 1])}`;
+}
+
+export function getPlatformString(keyCombinationString: string): string {
+    return splitKeyCombinationString(keyCombinationString)
+        .map(toPlatformString)
+        .join(", ");
+}
+
+function checkKey(event: KeyboardEvent, key: string): boolean {
+    return event.code === key;
+}
+
+function checkModifiers(
+    event: KeyboardEvent,
+    optionalModifiers: Modifier[],
+    activeModifiers: string[]
+): boolean {
+    return modifiers.reduce(
+        (matches: boolean, modifier: string, currentIndex: number): boolean =>
+            matches &&
+            (optionalModifiers.includes(modifier as Modifier) ||
+                event.getModifierState(platformModifiers[currentIndex]) ===
+                    activeModifiers.includes(modifier)),
+        true
+    );
+}
+
+function check(
+    event: KeyboardEvent,
+    optionalModifiers: Modifier[],
+    modifiersAndKey: string[]
+): boolean {
+    return (
+        checkKey(event, modifiersAndKey[modifiersAndKey.length - 1]) &&
+        checkModifiers(event, optionalModifiers, modifiersAndKey.slice(0, -1))
+    );
+}
+
+const shortcutTimeoutMs = 400;
```

Comment:
I initially thought multi-key shortcuts were not working, because I was not pressing them fast enough. The old Qt code will accept a shortcut without any timeout - is there a reason why you added a timeout here? If we do need one, maybe it needs to be a few seconds long?

Reply:
Hm, Qt is the only platform that I know that allows multi key shortcuts, so I don't really have anything to compare to. But making it a few seconds should be fine.

Reply:
Emacs is another example - it doesn't have a timeout either FWIW :-)